test: fix test failure in saucelabs ivy ie10

devversion opened the pull request:
One of the ivy acceptance tests currently fails in IE10. This is because we
recently added a new test that asserts that injecting `ViewRef` results in
a `NullInjectorError`.

Due to limitations in TypeScript and in polyfills for `setPrototypeOf`, the error
cannot be thrown as `ViewRef` is always considered injectable. In reality, `ViewRef`
should not be injectable, as explicitly noted in https://github.com/angular/angular/commit/c00f4ab2ae703e8a3ee4a3850170c9bbf0f9bc42.

There seems no way to simulate the proper prototype chain in such browsers
that do not natively support `__proto__`, so TypeScript and `core-js` polyfills
simply break the prototype chain and assign inherited properties directly
on `ViewRef`. i.e. so that `ViewRef.__NG_ELEMENT_ID__` exists and DI picks it up.

There is a way for TypeScript to theoretically generate proper prototype chain in
ES5 output, but they intend to only bother about the proper prototype chain in ES6
where `setPrototypeOf` etc. are officially standardized. See the response: https://github.com/microsoft/TypeScript/issues/1601#issuecomment-94892833.


I have a feeling  that it's better keeping the `hasOwnProperty` logic (as added with c00f4ab2ae703e8a3ee4a3850170c9bbf0f9bc42),
as IE10 will not fail but rather just be less strict. At the same time though, it could also
be an option to just enforce consistency by reverting your fix. Either way seems not 100%
ideal. I assume it's a conceptual issue with how `__NG_ELEMENT_ID` is handled.

AndrewKushnir commented:
Hi @devversion, thanks for fixing this problem 👍 

> I have a feeling that it's better keeping the hasOwnProperty logic (as added with c00f4ab),
> as IE10 will not fail but rather just be less strict.

Just wanted to mention that support for IE 9 and 10 will most likely be removed in v11, thus we should be able to rely on `hasOwnProperty` (since `setPrototypeOf` would be [available](https://caniuse.com/#search=setPrototypeOf) in all browsers that Angular officially supports) and keep the changes from c00f4ab.

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -2304,6 +2304,18 @@ describe('di', () => {
+    // not being able to simulate the prototype chain. This means that Angular's DI system
+    // considers `ViewRef` as injectable due to it having a `__NG_ELEMENT_ID__` directly
+    // assigned. We skip this test in such cases. This is currently the case in IE9 and
+    // IE11 as those don't support `__proto__`. Related TypeScript issue:
```
```suggestion
    // IE10 as those don't support `__proto__`. Related TypeScript issue:
```
I believe it should be IE10, since `setPrototypeOf` is supported in IE11, according to [canisue.com](https://caniuse.com/#search=setPrototypeOf).

devversion replied:
You are totally right! 😄 Just a typo!

devversion committed:
test: fix test failure in saucelabs ivy ie10

One of the ivy acceptance tests currently fails in IE10. This
is because we recently added a new test that asserts that injecting
`ViewRef` results in a `NullInjectorError`.

Due to limitations in TypeScript and in polyfills for `setPrototypeOf`,
the error cannot be thrown as `ViewRef` is always considered injectable.
In reality, `ViewRef` should not be injectable, as explicitly noted
in https://github.com/angular/angular/commit/c00f4ab2ae703e8a3ee4a3850170c9bbf0f9bc42.

There seems no way to simulate the proper prototype chain in such
browsers that do not natively support `__proto__`, so TypeScript
and `core-js` polyfills simply break the prototype chain and
assign inherited properties directly on `ViewRef`. i.e. so that
`ViewRef.__NG_ELEMENT_ID__` exists and DI picks it up.

There is a way for TypeScript to theoretically generate proper
prototype chain in ES5 output, but they intend to only bother
about the proper prototype chain in ES6 where `setPrototypeOf`
etc. are offically standarized. See the response:

https://github.com/microsoft/TypeScript/issues/1601#issuecomment-94892833.

devversion commented:
@AndrewKushnir Thanks for chiming in on this! I agree. Given that this removal is happening in the future, and that IE9/10 will just behave slightly less strict, it should be fine continuing to rely on `hasOwnProperty` while disabling this test for those less stricter browsers (i.e. the ones not supporting `__proto__` and `setPrototypeOf`).

crisbeto reviewed (approved):
LGTM
